### DISCUSSION TITLE
[Feature proposal] Add an alternate color palette for dark mode

### DISCUSSION THREAD

**Discussion by vitto** [tailwindlabs/tailwindcss#4130] [Ideas]
It would be nice to add the possibility to add a dark color palette theme, which can be switched with tailwind `dark-mode` config. 

### Why this request?

This whould be a killer feature for who wants to use the same contrast ratios for light and dark theme and gain in accessibility:

|  Contrast ratio  | Color | HEX   | Base  |
|---|---|---|---|
|   |   |   | 
| 14 | blue-01 | `#032D52` | `#FFFFFF` |
| 5.75 | blue-02 | `#0559A4` | `#FFFFFF` |
| 2.25 | blue-03 | `#3EA1F9` | `#FFFFFF` |
| 1.32 | blue-04 | `#B8DDFD` | `#FFFFFF` |
| 1.05 | blue-05 | `#F3F9FF` | `#FFFFFF` |
| 14 | blue-01 | `#AED8FD` | `#000000` |
| 5.75 | blue-02 | `#2B98F9` | `#000000` |
| 2.25 | blue-03 | `#055398` | `#000000` |
| 1.32 | blue-04 | `#02294A` | `#000000` |
| 1.05 | blue-05 | `#010911` | `#000000` |

Blue `#0559a3` with `#FFFFFF` color base [example](https://leonardocolor.io/?colorKeys=%230559a3&base=ffffff&ratios=1.05%2C1.32%2C2.25%2C5.75%2C14&mode=HSL).
Blue `#0559a3` with `#000000` color base [example](https://leonardocolor.io/?colorKeys=%230559a3&base=000000&ratios=1.05%2C1.32%2C2.25%2C5.75%2C14&mode=HSL).

### Dark mode color would bring these benefits

- Dark mode can be **enabled automatically**, reducing your codebase, without adding manual declarations in the markup, and you'd still use manual declarations to adjust small variations if needed;
- This would be nice for **accessibilty**, because the same contrast ratio means readable also for people with optical desease;
- This is a **true dark mode**, due to the same contrast ratio with light and dark theme colors.

These articles can be useful to explain why color would work better in this way:
[Creating contrast based themes with Leonardo](https://uxdesign.cc/creating-contrast-based-themes-with-leonardo-32b6219a090f)
[Leonardo: an open source contrast-based color generator](https://medium.com/@NateBaldwin/leonardo-an-open-source-contrast-based-color-generator-92d61b6521d2)

---

Just as basic example, in the confing it could be something like this:

```js
module.exports = {
  mode: 'jit',
  purge: [
    './src/**/*.{js,jsx}'
  ],
  darkMode: 'media',
  theme: {
    colors: {
      transparent: 'transparent',
      current: 'currentColor',
      blue: {
        '00': '#032D52',
        '02': '#0559A4',
        '03': '#3EA1F9',
        '04': '#B8DDFD',
        '05': '#F3F9FF',
      },
    },
    darkModeColors: {
      blue: {
        '00': '#AED8FD',
        '02': '#2B98F9',
        '03': '#055398',
        '04': '#02294A',
        '05': '#010911',
      },
    },
    extend: { },
  },
  variants: {
    extend: { },
  },
  plugins: [],
}
```

I cannot imagine myself without Tailwind in my stack, and I think this would be a very useful feature for it.

**Comment by boyankostov** [comment]
I was looking for the same feature. That would ideally allow me to not have to write separate dark: styles for each component. Hope it's planned for the future. It would allow building more intelligent dark palettes like the ones [IBM](https://carbondesignsystem.com/guidelines/color/overview/#dark-themes) and [Adobe](https://spectrum.adobe.com/page/color-palette/#Latest-Spectrum-colors) have, the latter one based on the same tool made by @NateBaldwinDesign mentioned above.

**Reply by wongjn** [reply]
Generally, the maintainers don't really recommend this approach and thus is probably why there is not a more "built-in" way of achieving this.

For example, from [Simon Vrachliotis](https://github.com/tailwindlabs/tailwindcss/discussions/3583#discussioncomment-368696):
> When implementing dark mode interfaces, you likely don't always want to just "swap" or invert colors, and often need much more "fine control", to make contextual decisions on a lot of parts of the UI.

And from [Adam Wathan](https://github.com/tailwindlabs/tailwindcss/discussions/10274#discussioncomment-4627634):
> Personally I think doing dark mode this way is seductive but not as practical as you expect. Tons of situations where you end up swapping the same color in light mode for two different colors in dark mode depending on how it’s used, and then you need to come up with different names for those two colors and it quickly becomes a very big project.

**Reply by vincerubinetti** [reply]
Quoting @simonswiss:

> you likely don't always want to just "swap" or invert colors, and often need much more "fine control"

I definitely agree, but...

First, to clarify, I don't think we want to "invert" in the traditional/literal sense of just inverting the RGB values (which always looks ugly). I'm imagining what the [Nightwind plugin](https://nightwindcss.com/) does. It "inverts" colors by swapping their TW palette brightness values, e.g. `bg-red-200` becomes to `bg-red-800`. Since the TW palette seems to be chosen to mostly equalize brightness across hues, I think it would give at least _better_ (not perfect) results on average than naive RGB inverting.

Also, just because we don't _always_ want to do something doesn't mean it can't be really useful _sometimes_. You're definitely going to need finer control at some point, but having a inverted palette could be really useful:

1) Using it as a _starting point_. It might be plenty good enough for a lot or even most cases, and can then just be overridden when needed.
2) Using it as a _fallback_. If I put a `dark:` variant on a `bg-` color for a tile component but forget to put the matching `text-` color, the component could become completely unusable and inaccessible. The Nightwind approach would seem to at least offer a better default.

> Here's a tweet by Steve Schoger that captures this really well: 
> > Close elements should still be lighter and distant elements should still be darker – even in a dark UI.
> > <img height="100" alt="image" src="https://github.com/user-attachments/assets/3c222a53-92d4-47b5-89d3-ba2c21b226da" />

This is a compelling argument, and I definitely agree we should use manual `dark:` variant wherever it's essential, like in that example. But again, I think there are plenty of cases (possibly even the majority?) where this type of inversion isn't an issue. When working with colors, we're not always aiming for Material-UI-esque "surfaces" meant to emulate real life.

---

Quoting @adamwathan:

> Personally I think doing dark mode this way is seductive but not as practical as you expect. Tons of situations where you end up swapping the same color in light mode for two different colors in dark mode depending on how it’s used

Definitely agree that there will be many situations where inverting isn't enough. I just think that perhaps the two approaches don't have to be mutually exclusive.

I could see something like this, for example:

```html
<body class="text-slate-900">
  <div class="bg-sky-100 dark:bg-indigo-900">lorem ipsum</div>
</div>
```

where the bg colors are specifically chosen for each mode, and the text color on `body` inherits down and automatically inverts to `text-slate-100` since no `dark:` variant is specified.

---

Maybe I'm in the minority, but this is how I wish it worked natively. But I could also see the maintainers wanting such a solution to live in plugin-space. Unfortunately Nightwind doesn't support TW v4, and may have been abandoned, so if anyone knows an alternative or an easy way to implement the plugin for myself, please comment.

A caveat to my comments here is that I haven't tried this inverting approach with the _Tailwind_ default color palette, but I have done it in fairly large projects with custom CSS var palettes (e.g. `--light-gray: #e0e0e0` becomes `--light-gray: #404040` in dark mode) with pretty good success.
